localstack/services/cloudformation/engine/transformers.py

```diff
@@ -50,27 +64,215 @@ def apply_transform_intrinsic_functions(
+        else:
+            processed_template = execute_macro(
+                account_id,
+                region_name,
+                parsed_template=template,
+                macro=transformation,
+                stack_parameters=stack_parameters,
+                transformation_parameters=transformation_parameters,
+            )
+
+    return processed_template
+
+
+def format_template_transformations_into_list(transforms: list | dict | str) -> list[dict]:
+    """
+    The value of the Transform attribute can be:
+     - a transformation name
+     - an object like {Name: transformation, Parameters:{}}
+     - a list a list of names of the transformations to apply
+     - a list of objects defining a transformation
+     so the objective of this function is to normalize the list of transformations to apply into a list of transformation objects
+    """
+    formatted_transformations = []
+    if isinstance(transforms, str):
+        formatted_transformations.append({"Name": transforms})
+
+    if isinstance(transforms, dict):
+        formatted_transformations.append(transforms)
+
+    if isinstance(transforms, list):
+        for transformation in transforms:
+            if isinstance(transformation, str):
+                formatted_transformations.append({"Name": transformation})
+            if isinstance(transformation, dict):
+                formatted_transformations.append(transformation)
+
+    return formatted_transformations
+
+
+def execute_macro(
+    account_id: str,
+    region_name: str,
+    parsed_template: dict,
+    macro: dict,
+    stack_parameters: dict,
+    transformation_parameters: dict,
+    is_intrinsic=False,
+) -> str:
+    macro_definition = get_cloudformation_store(account_id, region_name).macros.get(macro["Name"])
+    if not macro_definition:
+        raise FailedTransformationException(macro["Name"], "2DO")
+
+    formatted_stack_parameters = {}
+    for key, value in stack_parameters.items():
+        formatted_stack_parameters[key] = value.get("ParameterValue")
+
+    transformation_id = f"{account_id}::{macro['Name']}"
+    event = {
+        "region": region_name,
+        "accountId": account_id,
+        "fragment": parsed_template,
+        "transformId": transformation_id,
+        "params": transformation_parameters,
+        "requestId": long_uid(),
+        "templateParameterValues": formatted_stack_parameters,
+    }
+
+    client = connect_to(aws_access_key_id=account_id, region_name=region_name).lambda_
+    invocation = client.invoke(
+        FunctionName=macro_definition["FunctionName"], Payload=json.dumps(event)
+    )
+    if invocation.get("StatusCode") != 200 or invocation.get("FunctionError") == "Unhandled":
+        raise FailedTransformationException(
+            transformation=macro["Name"],
+            message=f"Received malformed response from transform {transformation_id}. Rollback requested by user.",
```

Comment:
I know we had them in before as well but we need to find a better abstraction in the future here since `Rollback requested by users` won't always be added here.